components/skus/browser/rs/lib/src/sdk/credentials/fetch.rs

```diff
@@ -173,11 +306,83 @@ where
 
         let resp: Vec<ItemCredentialsResponse> = serde_json::from_slice(resp.body())?;
+        event!(Level::DEBUG, "done decoded body");
 
         let mut time_limited_creds: HashMap<String, Vec<TimeLimitedCredential>> = HashMap::new();
+        let mut time_limited_v2_creds: Vec::<String> = Vec::new();
 
-        for item_cred in resp {
+        'outer: for item_cred in resp {
             match item_cred {
+                ItemCredentialsResponse::TimeLimitedV2 {
+                    item_id,
+                    blinded_creds,
+                    signed_creds,
+                    batch_proof,
+                    public_key,
+                    issuer_id,
+                    valid_from,
+                    valid_to,
+                    ..
+                } => {
+                    if let Some(item_creds) =
+                        self.client.get_time_limited_v2_creds(&item_id).await?
+                    {
+                        // ignore
+                        let from = NaiveDateTime::parse_from_str(&valid_from, "%Y-%m-%dT%H:%M:%SZ")
+                                    .map_err(|_| {
+                                        InternalError::InvalidResponse(
+                                            "Could not parse valid from".to_string(),
+                                        )
+                                    })?;
+                        let to = NaiveDateTime::parse_from_str(&valid_to, "%Y-%m-%dT%H:%M:%SZ")
+                                    .map_err(|_| {
+                                        InternalError::InvalidResponse(
+                                            "Could not parse valid to".to_string(),
+                                        )
+                                    })?;
+                        for stored_cred in item_creds.creds.unwrap_or(Vec::<TimeLimitedV2Credential>::new()) {
+                            if to == stored_cred.valid_to && from == stored_cred.valid_from {
+                                // we have already performed the batch proof of this credential
+                                continue 'outer;
+                            }
+                        }
+
+                        // so we can clear out the used tokens at the end of the loop
+                        time_limited_v2_creds.push(item_id.to_string());
+
+                        // Rederive blinded creds so that the proof will fail if different creds were signed
+                        let my_blinded_creds: Vec<BlindedToken> =
+                            item_creds.tokens.iter().map(|t| t.blind()).collect();
+
+                        // okay, right here we need to filter out all of the
+                        // creds that do not have a blinded token that match
+                        // prior to the batch proof check.
+                        let mut bucket_blinded_creds: Vec<BlindedToken> = Vec::new();
+
+                        for sbc in blinded_creds {
```

Comment:
I might use https://doc.rust-lang.org/std/collections/hash_set/struct.Intersection.html for this

Reply:
i think we need to implement Eq/Hash for challenge_bypass_ristretto::voprf::BlindedToken for this to work correctly?

Reply:
okay bummer, let's not block on that